Gettext support

MichaelHoste opened the pull request:
Hey @armandsar, I added Gettext support to this package.

A lot of Gettext syntaxes are now supported. You can see how in that file: `tests/fixtures/gettext/example.php`.

As you can see, some are quite classic (`_("")` or `_n("", "", number)`) and some are added for easier interpolation (`_i("hello %s", "world")`).

There are some tests for both client/server behaviour but also to test GetText functions.

You'll see that there are also `sync_and_purge` and `sync_and_show_purgeable` methods in the package. They work like to the ones in Ruby: https://github.com/aurels/translation-gem#usage

I also renamed `php artisan translations:xxx` to `php artisan translation:xxx` for better consistency, I hope you don't mind.

Please tell me if you see something weird in the package, I certainly did some mistakes.

Also, to change the language in your Laravel application, you need to do:

```php
$t = new TranslationIO(config('translationio'));
$t->setLocale('fr');
```

Do you know if there is a way to do something like `TranslationIO::setLocale('fr')` instead? I think I saw something like this somewhere but I'm not very good with Facades and PHP meta-programming.

MichaelHoste committed:
First working Gettext-compatible version

MichaelHoste committed:
Merge branch 'master' into gettext

MichaelHoste committed:
Class renaming and better translator_functions

MichaelHoste committed:
Gettext methods are working, Gettext tests with VCR and evaluation of gettext translations

MichaelHoste committed:
fix YAML multi-line

MichaelHoste committed:
Add sync_and_purge and sync_and_show_purgeable commands + better stdout, rename translations: to translation: in commands for better consistency

armandsar commented on src/Console/Commands/Init.php:
```diff
@@ -26,7 +25,6 @@ public function __construct(InitService $service)
         $this->info('Init started');
-        $this->service->call();
-        $this->info('Init finished');
+        $this->service->call($this);
```
Could use https://github.com/wilderborn/partyline instead of passing command object

MichaelHoste replied:
I just tried to do this by creating AbstractCommand (like in the README). But I've got `ErrorException: Non-static method Wilderborn\Partyline\Partyline::bind() should not be called statically` when Partyline is called and I don't know how to include it into our package.

(I guess it's still something related to Facades to include in our own ServiceProvider, right?)

armandsar replied:
import at the top with:
"use Wilderborn\Partyline\Facade as Partyline"

armandsar committed:
add whitelist to phpunit

armandsar commented on src/GettextPOGenerator.php:
```diff
@@ -0,0 +1,186 @@
+        $this->config = $application['config']['translationio'];
+    }
+
+    public function call($source, $targets)
```
source is not needed?

armandsar commented on src/GettextTranslationSaver.php:
```diff
@@ -0,0 +1,62 @@
+        $moFile = $lcMessagesDir . DIRECTORY_SEPARATOR . 'app.mo';
+
+        if ( ! $this->filesystem->exists($localeDir)) {
+            $this->filesystem->makeDirectory($localeDir, 0777, true);
```
Could pass 4th argument as true, to force creating of directory, wont need to check if it exists

armandsar commented on src/Service/Sync.php:
```diff
@@ -87,6 +130,92 @@ private function makeRequest($client, $body)
+    private function displayUnusedSegments($responseData, $command, $showPurgeable, $purge) {
+        $unusedSegments = collect($responseData['unused_segments']);
+
+        $yamlUnusedSegments    = $unusedSegments->filter(function ($unusedSegment) {
```
Its very uncommon to align assignments like this in php.
Actually you can run " vendor/bin/phpcbf --standard=psr2 src/", thats probably simplest way to get "most common" formating

armandsar commented on src/ServiceProvider.php:
```diff
@@ -15,23 +20,30 @@ class ServiceProvider extends LaravelServiceProvider
+
+        $translationIO = new TranslationIO($config);
+
+        $translationIO->setLocale($config['source_locale']);
```
!! Dont think we can do this here, this should be done by user, depending on their add logic

MichaelHoste replied:
We did it like this in Rails and it doesn't seem to bother anyone. You always need a default language and the source locale seems a good fit.

In the code, the best way to manage locales is to add setLocale at controller's level (using session, cookie, params in URL, subdomain or anything else).

In Rails this is the code we suggest to users and it will always override the default locale: https://github.com/aurels/translation-gem/blob/8ccb7df8b39248d2f408d93d67ee72651f7bccbf/lib/translation_io/controller.rb#L3

armandsar replied:
You are right, cant explain why I though it could cause any problems..

armandsar commented on src/ServiceProvider.php:
```diff
@@ -47,8 +59,19 @@ private function configureCommands()
+     */
+    public static function improveGettextFunctions()
+    {
+        include_once __DIR__.'/translator_functions.php';
```
We can instruct composer to autoload helpers for us, see example here:
https://github.com/armandsar/laravel-quick-view/blob/master/composer.json#L30

armandsar commented on src/translator_functions.php:
```diff
@@ -0,0 +1,192 @@
+ * Returns the translation of a string with interpolation.
+ */
+if (!function_exists('i_')) {
+    function i_($original)
```
Maybe its worth proxying these function to some class? Especially given that few of them have the same body.

MichaelHoste replied:
I tried but actually only 2 of them have the same body (`i_` and `i__`) and since there is strange things happening using `func_get_args()`, it's already more complex than just doing something like: 

```php
function i__($original)
{
  return i_($original);
}
```

because it will lose the extra (hidden) arguments.

I have to admit I don't care a lot about this specific duplication of code. That code will not evolve much and it's just 10 lines of code.

armandsar committed:
add Facade

armandsar commented:
Hey, I have added Facade and a few comments. Nothing too crazy :)
Just need to check logic for setting app locale. 
`LaravelApp::setLocale(substr($locale, 0, 2));` - this line is commented, but this actually would be correct for an app using both gettext and key values, right? Because laravel only uses "basic" locales.

Have you tried this on a real application? I see that many parts of sync/save logic dont have test (which is fine now). I am just thinking - if you say that its working for you and you wont be making any more changes to this feature, I might take if for refactoring and writing tests (unless you really want to do it yourself).

Btw, a test is failing.

MichaelHoste commented:
Thanks a lot for your help @armandsar!

I'll have a look at this later this week. Your comments make a lot of sense, I'll try to improve the code and make the tests pass (they work locally and the package already work on a real project, I don't know why this test breaks :-/ )

armandsar commented:
@MichaelHoste that test fails for me as too, are you sure it passes for you now?
Or maybe its related to your comments -  "# IMPORTANT: locale must be installed in server!".
Could you clarify that?

MichaelHoste commented:
@armandsar I must admit this is some code I copy-pasted from here: https://github.com/eusonlito/laravel-Gettext/blob/master/src/Eusonlito/LaravelGettext/Gettext.php#L155

So I'm not sure how it behaves... It's still working on my side so my guess is that the locale is maybe not installed on your machine (and on the CI server). I checked and on MacOS there is already a lot of pre-installed locales (`locale -a` on console returns 203 languages). That should be why it's already working on my side.

Are you on Linux? Maybe this link will do the trick: https://askubuntu.com/questions/76013/how-do-i-add-locale-to-ubuntu-server

This is somewhat confirmed there: http://www.php.net/manual/en/book.gettext.php#91187

If this works, we'll need to specify this in the README/installation part.

Thanks for your feedback! It was not obvious the failed test was related to that!

armandsar committed:
fix orders of arguments in test assertions

armands2ar commented:
@MichaelHoste indeed I didnt have any locales besides EN on my machine.
Did install lv with "sudo apt-get install language-pack-lv"
`locale -a | grep lv`
gives
`lv_LV.utf8`

But in the code we are setting it to "lv.UTF-8" and changing test to 
`Facade::setLocale('lv_LV');`
makes it pass.

Not sure what would be best approach to solving this.. 
Did you have this kind of issue in ruby gem?
Will be waiting for your thoughts on this.

armandsar commented:
@MichaelHoste indeed I didnt have any locales besides EN on my machine.
Did install lv with "sudo apt-get install language-pack-lv"
`locale -a | grep lv`
gives
`lv_LV.utf8`

But in the code we are setting it to "lv.UTF-8" and changing test to 
`Facade::setLocale('lv_LV');`
makes it pass.

Not sure what would be best approach to solving this.. 
Did you have this kind of issue in ruby gem?
Will be waiting for your thoughts on this.

MichaelHoste commented:
@armandsar in Ruby we don't have this issue because Gettext is not part of the language but is implemented using the language: https://github.com/ruby-gettext/gettext

I think https://github.com/oscarotero/Gettext is an PHP alternative to do this but it uses the same syntax than Laravel (`__`). And the `_` syntax will be assigned to the original GetText anyway. So that makes things harder for us.

I would prefer to use the original PHP Gettext. I think the solution for us in this situation is:

 * To notify the user when he translation:sync (or start the app) and a locale is missing on the system (with instructions how to add it, it's easy enough).
 * To fallback xx to xx_XX if the locale is not found (or to restrict the available locales to xx_XX on Translation.io for a Laravel project). Because both exists:

<img width="370" alt="capture d ecran 2018-02-27 a 09 55 07" src="https://user-images.githubusercontent.com/529357/36719078-55d59554-1ba4-11e8-8177-68a7018eff0d.png">

> laravel only uses "basic" locales.

It's another issue I have with Laravel. It's kind of weird because `zh-Hant`, `zh-Hans` (traditional and simplified Chinese) are very different languages. How do you deal with them using the standard Laravel localization? Are you sure only 2 letters are accepted? I tried to find a list of all available locales but without success.

armandsar commented:
@MichaelHoste 
> laravel only uses "basic" locales.

Thats actually wrong, right thing to say would be that its common way to localize.
Laravel actually does not care what locale is. "Foo" works just as good as "lv_LV" or "en".

So I guess this isnt that big of an issue. We just need to specify in readme that if people want to use gettext they have to use locales with region.

> To notify the user when he translation:sync (or start the app) and a locale is missing on the system (with instructions how to add it, it's easy enough).

Given that translations will be synced from dev machines, this wont help with using gettext on production, but again an info in readme to pointed out locale required on server might be enough.

armandsar committed:
add using full locale name for test

armandsar committed:
add language packs to travis

armandsar commented:
Added language packs to travis, test should pass now

MichaelHoste committed:
remove unused argument, force makeDirectory

MichaelHoste committed:
typo

MichaelHoste committed:
autoload translator_functions in composer

MichaelHoste commented:
Hi @armandsar,

Thanks for adding the language packages on Travis!

I did some small stuff on the code : 

 * Remove unused argument ($source).
 * Force directory creation like you suggested
 * Autoload translator_functions directly from composer.json

Tomorrow I'll try to work on the missing locale issue (trigger warnings etc.) and write a better README (one like https://github.com/aurels/translation-gem) at the same time if that's ok with you.

armandsar commented:
That sounds great!

armandsar committed:
change phpunit dependency

MichaelHoste commented:
The more I think about it and the more I'm enclined to NOT use the native PHP gettext extension for the following reasons:

 * There is the system locale issue that needs to be installed and adds some friction.
 * Gettext has an aggressive caching system that can sometimes be flushed only by restarting the server. It can be a real pain in development (it can be bypassed with symlinks hacks but that's not clean).
 * If I'm not wrong, since Gettext depends on environment variables, it's not thread-safe. And we don't want to have language race-conditions between connected users in situations with multi-threaded servers (is it common in the PHP-Laravel-world?)

I'll try to use `use Gettext\Translator;` instead of `use Gettext\GettextTranslator;` ([here](https://github.com/oscarotero/Gettext#usage-example)).

The trade-off would be that classic gettext methods (`_`, `_n` etc.) would not be usable anymore. And it's a shame. But we can still use `i__`, `n__`, `p__` and `np__` (even if that's really ugly)

What do you think about this?

MichaelHoste committed:
replace php-gettext by oscarotero/Gettext

New approach for easier installation, better cache invalidation and
thread-safe Gettext, better API versioning.

Only 4 gettext keywords : __t, __n, __p and __np

MichaelHoste commented:
@armandsar I just pushed a new version that uses [oscarotero/Gettext](https://github.com/oscarotero/Gettext) instead of php-gettext to be able to use the gettext methods and read the MO files.

It provides an easier installation (no need to install system languages anymore), a better cache invalidation and a thread-safe gettext (I'm not sure about that last one).

Only 4 gettext keywords are now available: `__t`, `__n`, `__p` and `__np`. It will not be possible to use the default Gettext methods but I'm not sure people really care.

I got 2 questions for you, I don't know if you can help:

 * Do you have any idea why the "prefer-lowest" test fails? It seems related to TestBench and PHPUnit but I don't know how.
 * I still have some trouble using `Partyline` even using `"use Wilderborn\Partyline\Facade as Partyline"`. I'm not sure what I'm missing here (I agree with you that it would look better, but it's not a disaster if we can't use it ^^).

--------

Later this week I'll try to improve the "init" to upload the existing GetText translations that are already in the project.

--------

I hope this PR will soon be able to get merged :-)

armandsar committed:
update phpunit version

armandsar committed:
update travis composer

armandsar commented:
@MichaelHoste ignore that test, I will see later what can be done.
And, yes, Partyline is whatever :)

> Later this week I'll try to improve the "init" to upload the existing GetText translations that are already in the project.

As part of this PR? Otherwise I guess I could merge

MichaelHoste commented:
> As part of this PR? Otherwise I guess I could merge

@armandsar I think I could be able to do it tomorrow, so it's best to wait a bit more ;-)

MichaelHoste committed:
allow init with existing PO

MichaelHoste commented:
@armandsar If you agree to the last changes, this branch can finally be merged :-)

Since the "init" now uploads existing po translations, we can clone a translation.io project by creating a new project, copying the new api key in the config file and doing an "init" again. It will create exactly the same project.

It's sometimes useful to recreate a project from scratch if something went wrong during translation of merging of a branch.

Or if we want to branch the project for another client.

MichaelHoste committed:
wrong use of cloned variable

armandsar commented:
@MichaelHoste seems fine :+1:
